azure-functions-powershell-library/WorkerIndexingHelper.cs

```diff
@@ -0,0 +1,197 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using AzureFunctions.PowerShell.SDK;
+using System.Collections.ObjectModel;
+using System.Management.Automation.Language;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker
+{
+    internal class WorkerIndexingHelper
+    {
+        internal static List<FunctionInformation> IndexFunctions(string baseDir)
+        {
+            if (!Directory.Exists(baseDir))
+            {
+                throw new FileNotFoundException();
+            }
+            List<FileInfo> powerShellFiles = GetPowerShellFiles(Directory.CreateDirectory(baseDir));
+            List<FunctionInformation> rpcFunctionMetadatas = new List<FunctionInformation>();
+
+            foreach (FileInfo powerShellFile in powerShellFiles)
+            {
+                rpcFunctionMetadatas.AddRange(IndexFunctionsInFile(powerShellFile));
+            }
+
+            return rpcFunctionMetadatas;
+        }
+
+        private static IEnumerable<FunctionInformation> IndexFunctionsInFile(FileInfo powerShellFile)
+        {
+            List<FunctionInformation> fileFunctions = new List<FunctionInformation>();
+            var fileAst = Parser.ParseFile(powerShellFile.FullName, out _, out ParseError[] errors);
+            if (errors.Any())
+            {
+                throw new Exception($"Couldn't parse the file: {powerShellFile.FullName}");
+                // TODO: Probably don't throw here?
+                //return fileFunctions;
+            }
+            if (powerShellFile.Extension == ".ps1") 
+            {
+                // parse only the file param block, return one RpcFunctionMetadata assuming the file is the entry point
+                var paramAsts = fileAst.ParamBlock;
+                if (paramAsts != null && paramAsts.Attributes.Where(x => x.TypeName.ToString() == "Function").Any()) 
+                {
+                    // This is a function, return it 
+                    fileFunctions.Add(CreateRpcMetadataFromFile(powerShellFile.FullName));
+                }
```

Comment:
Can you please clarify this assumption regarding entry points? Not sure I follow

Reply:
In a .ps1 file, the file itself has a param block and is treated as the function, meaning the worker is not calling any function defined in the file to invoke the Function. In .psm1 files, any function defined in the file may be a Function if it has the correct attribute.